File: hil/source/hil/tools/test_device_identity.m
function result=test_device_identity()
% Test exact device identity parsing and reject invalid configurations.
cleanup=gpenmpc_test_device_config(); %#ok<NASGU>
file=getenv('GPENMPC_DEVICE_CONFIG');fixture=jsondecode(fileread(file));
actual=gpenmpc_device_identity();
expected=bitor(bitshift(uint64(hex2dec('11223344')),32),uint64(hex2dec('55667788')));
assert(strcmp(actual.uid,fixture.uid)&&actual.uid_uint64==expected);
bad={struct(),setfield(fixture,'uid',123),setfield(fixture,'uid','0'), ...
    setfield(fixture,'uid','18446744073709551616'),setfield(fixture,'uid','*'), ...
    setfield(fixture,'px4_guid',repmat('A',1,36)), ...
    setfield(fixture,'bootloader_sn_display',repmat('0',1,24))}; %#ok<SFLD>
for k=1:numel(bad)
    f=fopen(file,'w');assert(f>=0);fprintf(f,'%s',jsonencode(bad{k}));fclose(f);
    rejected=false;
    try,gpenmpc_device_identity();catch ex,rejected=strcmp(ex.identifier,'GPENMPC:DeviceConfiguration');end
    assert(rejected,'Invalid device configuration was accepted.');
end
setenv('GPENMPC_DEVICE_CONFIG',fullfile(fileparts(file),'missing.json'));
rejected=false;
try,gpenmpc_device_identity();catch ex,rejected=strcmp(ex.identifier,'GPENMPC:DeviceConfiguration');end
assert(rejected,'Missing device configuration was accepted.');
result=struct('passed',true,'exact_uint64',true,'invalid_configurations',numel(bad)+1);
disp(result);
end

File: hil/source/hil/tools/test_device_identity.py
"""Test device configuration parsing without device access."""
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile

from gpenmpc_device_identity import device_identity


def main():
    fixture = dict(uid='1234605616436508552', px4_guid='000600000000111111112222222233333333',
                   bootloader_sn_display='333333332222222211111111')
    prior = os.environ.get('GPENMPC_DEVICE_CONFIG')
    with tempfile.TemporaryDirectory(prefix='gpenmpc_identity_') as folder:
        path = Path(folder) / 'device.json'
        os.environ['GPENMPC_DEVICE_CONFIG'] = str(path)
        cases = [None, {}, dict(fixture, uid=123), dict(fixture, uid='0'),
                 dict(fixture, uid='18446744073709551616'), dict(fixture, uid='*'),
                 dict(fixture, px4_guid='A' * 36), dict(fixture, bootloader_sn_display='0' * 24)]
        try:
            for case in cases:
                if case is not None:
                    path.write_text(json.dumps(case), encoding='utf-8')
                try:
                    device_identity()
                except (ValueError, FileNotFoundError):
                    pass
                else:
                    raise AssertionError('Invalid identity accepted')
            path.write_text(json.dumps(fixture), encoding='utf-8')
            assert device_identity() == fixture
            assert device_identity('uid') == fixture['uid']
            env = dict(os.environ)
            for key in ('GPENMPC_DEVICE_CONFIG', 'GPENMPC_EXTERNAL_PATHS', 'GPENMPC_RECOVERY_ROOT',
                        'HIL_RECOVERY_RECORDS', 'GPENMPC_RFLY_ROOT'):
                env.pop(key, None)
            source = Path(__file__).resolve().parent
            code = ('import sys; import gpenmpc_application_upload_once as u; '
                    'u.read_image("fmuv6c"); '
                    'assert "serial" not in sys.modules and "px_uploader" not in sys.modules')
            result = subprocess.run([sys.executable, '-B', '-c', code], cwd=source, env=env,
                                    capture_output=True, text=True, timeout=30)
            assert result.returncode == 0, result.stderr
            result = subprocess.run([sys.executable, '-B', str(source/'gpenmpc_application_upload_once.py'), '--help'],
                                    env=env, capture_output=True, text=True, timeout=30)
            assert result.returncode == 0, result.stderr
        finally:
            if prior is None:
                os.environ.pop('GPENMPC_DEVICE_CONFIG', None)
            else:
                os.environ['GPENMPC_DEVICE_CONFIG'] = prior
    print(json.dumps(dict(passed=True, invalid_configurations=len(cases), exact_identity=True,
                         import_without_private_records=True, help_without_private_records=True)))


if __name__ == '__main__':
    main()
